tests/integration/utils.py

```diff
@@ -0,0 +1,187 @@
+        basic_auth = config['auth']['basic_auth']
+        session.auth = (basic_auth['username'], basic_auth['password'])
+        session.verify = False
+    else:
+        oauth2 = config['auth']['oauth2']
+        data = {
+            'client_id': oauth2['client_id'],
+            'client_secret': oauth2['client_secret'],
+            'grant_type': 'client_credentials'
+        }
+        try:
+            res = requests.post(url=oauth2['token_api_url'], data=data)
+            token = res.json()['access_token']
+        except KeyError:
+            sys.exit('Error: invalid OAuth2 credentials')
+        session.headers = {'Authorization': f'Bearer {token}'}
+
+    return session
+
+
+# Get response content in JSON format
+def get_json_content(self, response):
+    try:
+        return response.json()
+    except json.decoder.JSONDecodeError:
+        self.fail('Response not in JSON format')
+
+
+# Get resource schema from OpenAPI specification
+def get_resource_schema(self, resource):
+    return self.openapi['definitions'][resource]['properties']
+
+
+# Helper function to make a web request and lightly validate the response
+def make_request(self, endpoint, expected_status_code,
+                 params=None,
+                 max_elapsed_seconds=5):
+    """
+    Keyword arguments:
+    * endpoint -- the endpoint to request
+    * expected_status_code -- expected HTTP status code
+    * params -- [optional] key-value pairs parameters (default: None)
+    * max_elapsed_seconds -- [optional] maximum elapsed times (default: 5)
+    Return:
+    A response object contains a server’s response to an HTTP request
+    """
+
+    requested_url = f'{self.base_url}{endpoint}'
+    response = self.session.get(requested_url, params=params)
+    logging.debug(f'Sent request to {requested_url}')
+    # Response status code should be as expected
+    status_code = response.status_code
+    self.assertEqual(status_code, expected_status_code)
+
+    # Response time should less then max_elapsed_seconds
+    elapsed_seconds = response.elapsed.total_seconds()
+    logging.debug(f'Request took {elapsed_seconds} second(s)')
+    self.assertLess(elapsed_seconds, max_elapsed_seconds)
+
+    return response
+
+
+# Check the schema of response match OpenAPI specification
+def check_schema(self, response, schema):
+    # Helper function to get attributes of the schema
+    def __get_schema_attributes():
+        return schema['attributes']['properties']
+
+    # Helper function to map between OpenAPI data types and python data types
+    def __get_attribute_type(attribute):
+        types_dict = {
+            'string': str,
+            'integer': int,
+            'int32': int,
+            'int64': int,
+            'float': float,
+            'double': float,
+            'boolean': bool,
+            'array': list,
+            'object': dict
+        }
+        if '$ref' in attribute:
+            openapi_type = __get_object_type(attribute['$ref'])
```

Comment:
Nice work adding logic here to parse refs in the swagger document. Currently, won't this only parse the value of `type` though? I think this should follow the same logic as `__get_attribute_type()` and first attempt to parse `format` and if the field isn't found or the value doesn't exist in `types_dict`, the `type` can be used instead. Once this is implemented, I think this could be added to the skeleton.

Reply:
I showed this to @tsoliangwu0130 and he agrees that this would be great to add to the skeleton. He also brought up the case of a ref that references another ref. I think the logic could be modified to allow these kinds of refs. There's also the case of circular refs, for example when one ref references another ref, which references the original ref. We would want to make sure that the script doesn't crash or hang indefinitely if it encounters a circular ref.